src/model/change-set/apply-changes.ts

```diff
@@ -0,0 +1,68 @@
+import { Project } from '../../project/project';
+import { ProjectSource } from '../../project/source';
+import { ChangeSet, TextChange } from './change-set';
+
+export class InvalidChangeSetError extends Error {
+    constructor(message: string) {
+        super(message);
+        this.name = this.constructor.name;
+    }
+}
+
+export function applyChangeSet(project: Project, changeSet: ChangeSet): Project {
+    const newSources = project.sources.map((source) => {
+        const changes = changeSet.changes.filter((c) => c.source.name === source.name);
+        return applyChanges(source, changes);
+    });
+
+    return new Project({
+        ...project,
+        sources: newSources,
+    });
+}
+
+export function applyChanges(
+    source: ProjectSource,
+    changes: ReadonlyArray<TextChange>,
+): ProjectSource {
```

Comment:
Since the function could also be used standalone it might make sense to add a check for each `TextChange` that it targets the correct file

Reply:
in my WIP commit that introduces the `AppendChange` in addition to the `TextChange`, this function will no longer be exported anyway.